add an option for url types so that it can be opened in current tab (master branch)

fbaligand opened the pull request:
In Management tab > Index Patterns, in field edition form with "URL" format and "Link" type, this PR adds a checkbox to indicate that this link is to be open in current tab (default is "in another tab", ie. `_blank`).

This is particularly useful when your link points to a kibana dashboard with filtered params.

Fix #12668

![image](https://user-images.githubusercontent.com/3750300/28754822-b15da734-754d-11e7-8f64-f39ccc86d40d.png)

elasticmachine commented:
Can one of the admins verify this patch?

fbaligand commented:
@epixa 
Could you review this PR ?

epixa commented:
@fbaligand Sorry for the lack of response, I was on vacation last week and am still catching up on notifications.  I'm sure we can get a review on this soon.

epixa commented:
jenkins, test this

chrisronline commented on src/ui/public/field_format_editor/editors/url/url.html:
```diff
@@ -7,6 +7,13 @@
 
+<div class="checkbox" ng-if="editor.formatParams.type == 'a'">
+  <label>
+    <input ng-model="editor.formatParams.openLinkInCurrentTab" class="ng-pristine ng-valid ng-touched" type="checkbox">
```
There really isn't a reason to manually add these classes (`ng-pristine`, `ng-valid` and `ng-touched`). Angular will handle that automatically.

BigFunger replied:
These attributes should be one per line. See [here](https://github.com/fbaligand/kibana/blob/master/style_guides/html_style_guide.md#multiple-attribute-values) for details.

epixa commented:
It looks like there are some linting issues caught by CI as well.

BigFunger commented on src/core_plugins/kibana/common/field_formats/types/url.js:
```diff
@@ -106,7 +106,9 @@ UrlFormat.prototype._convert = {
         }
 
-        return `<a href="${url}" target="_blank">${linkLabel}</a>`;
+        let linkTarget = this.param('openLinkInCurrentTab') ? '_self' : '_blank';
```
Const that up. :)

BigFunger commented on src/ui/public/field_format_editor/editors/url/url.html:
```diff
@@ -119,6 +126,7 @@ <h4 class="hintbox-heading">
   </div>
 
   <input ng-model="editor.formatParams.labelTemplate" class="form-control">
+
```
Remove the extra whitespace, please.

chrisronline commented:
Good catch.

Copied from CI (which is linked below):
```
 /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-intake/src/core_plugins/kibana/common/field_formats/types/url.js
  109:13  error  'linkTarget' is never reassigned. Use 'const' instead  prefer-const
  110:1   error  Trailing spaces not allowed                            no-trailing-spaces
```

fbaligand commented:
Thanks for your feedback and your review !
 I will fix all your remarks as soon as possible !

fbaligand commented:
I'm currently on holidays (my turn ^^). 
So I will fix remarks in about one week. 
Sorry for the delay...

fbaligand commented:
@epixa @chrisronline @BigFunger 
I just fixed all your remarks and rebased my code to last master version.

=> Could you review it ?

If you wish I squash my 2 commits, tell me.

chrisronline commented on src/ui/public/field_format_editor/editors/url/url.html:
```diff
@@ -8,6 +8,16 @@
 </div>
 
+<div class="checkbox" ng-if="editor.formatParams.type == 'a'">
+  <label>
```
@cjcenizal @timroes Any accessibility concerns with this label/input approach?

timroes replied:
@chrisronline nope, looks very fine :-) Thanks for pinging.

fbaligand commented:
@chrisronline 
For information, to write my html for checkbox, I used exactly the same style than this existing one :
https://github.com/elastic/kibana/blob/5514f343549509d1fa6be5fe2c1ed4cfb806356a/src/core_plugins/console/public/src/directives/settings.html#L34

chrisronline commented on src/ui/public/field_format_editor/editors/url/url.html:
```diff
@@ -8,6 +8,16 @@
+  <label>
+    <input
+      type="checkbox"
+      ng-model="editor.formatParams.openLinkInCurrentTab">
```
Can we move the `>` to the next line? Along the lines of [these examples](https://github.com/elastic/kibana/blob/master/style_guides/html_style_guide.md#multiple-attribute-values)

cjcenizal commented on src/ui/public/field_format_editor/editors/url/url.html:
```diff
@@ -8,6 +8,16 @@
   </select>
 </div>
 
+<div class="checkbox" ng-if="editor.formatParams.type == 'a'">
```
Would you mind changing this to `class="kuiCheckBox"`? This is the class provided by the [UI Framework[(https://github.com/elastic/kibana/tree/master/ui_framework).

fbaligand commented:
OK.
I just pushed a commit which fixes the checkbox html.

fbaligand commented:
@chrisronline @BigFunger 
Could you verify that your requested changes are done, and if so, give go for LGTM ?

chrisronline commented:
LGTM!

fbaligand commented:
Thanks @chrisronline !
Great news !

timroes commented on src/core_plugins/kibana/common/field_formats/types/url.js:
```diff
@@ -106,7 +106,9 @@ UrlFormat.prototype._convert = {
-        return `<a href="${url}" target="_blank">${linkLabel}</a>`;
+        const linkTarget = this.param('openLinkInCurrentTab') ? '_self' : '_blank';
+
+        return `<a href="${url}" target="${linkTarget}">${linkLabel}</a>`;
```
I think we should add [`rel=noopener`](https://mathiasbynens.github.io/rel-noopener/) since this may contain external links.

fbaligand commented:
@timroes 
Done !
=> Could you check it ?

timroes commented:
LGTM @BigFunger would you care for an approval (or dismiss of your review :D)?

timroes commented:
Jenkins, test this

timroes commented:
Sorry, I think adding that `rel=noopener` broke now 3 tests in [this test](https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/common/field_formats/types/__tests__/url.js) (line 10, 24 and 37 - also see [jenkins failure](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request+multijob-intake/10650/console)). Could you add the `rel=noopener` there, too?

fbaligand commented:
@timroes 
Sorry for these broken tests.
I fixed them.

timroes commented:
Jenkins, Test this

timroes commented:
Jenkins, test this

fbaligand commented:
@timroes 
Jenkins says it's OK.
So now, what's the next step ?
Should we wait for @BigFunger review ?

fbaligand commented:
Great @BigFunger !

fbaligand commented:
@timroes @BigFunger @chrisronline 
Given that all checks are green, 3 reviewers validated the PR, could you merge the PR ?

BigFunger commented:
6.x/6.1: 5c93c1fdedd303acbef2037f01c4fb2b8dc899b7
6.0: 7f85542d82219ba30b406d5dbdc972c3192e8692

fbaligand commented:
Houhou ! Great ! 
Thanks @BigFunger !

epixa commented:
Thanks a bunch for this contribution @fbaligand!

fbaligand commented:
@epixa happy to contribute !
With this PR, I now have contributed to Logstash, Elasticsearch and Kibana !

epixa commented:
@fbaligand Nice!

fbaligand commented:
By the way, I have another kibana PR ready to push !
A really nice feature up to me !
Teasing... :)
